drivers/eeprom/eeprom_fm25xxx.c

```diff
@@ -0,0 +1,266 @@
+{
+	if (size <= 500) {
+		return 1;
+	} else if (size <= 64000) {
+		return 2;
+	} else {
+		return 3;
+	}
+}
+
+static int eeprom_fm25xxx_set_enable_write(const struct device *dev, bool enable_writes)
+{
+	const struct fm25xxx_config *config = dev->config;
+	uint8_t op = enable_writes ? FM25XXX_WREN : FM25XXX_WRDI;
+
+	const struct spi_buf tx_bufs[] = {{
+		.buf = &op,
+		.len = sizeof(op),
+	}};
+	const struct spi_buf_set tx_buf_set = {
+		.buffers = tx_bufs,
+		.count = ARRAY_SIZE(tx_bufs),
+	};
+
+	spi_write_dt(&config->spi, &tx_buf_set);
+
+	return 0;
+}
+
+int eeprom_fm25xxx_read(const struct device *dev, off_t offset, void *data, size_t len)
+{
+	int ret;
+	const struct fm25xxx_config *config = dev->config;
+	struct fm25xxx_data *dev_data = dev->data;
+
+	if (offset + len > config->size) {
+		LOG_ERR("Can not read more data than the device size");
+		return -EINVAL;
+	}
+
+	if (len == 0) {
+		return 0;
+	}
+
+	uint8_t read_op[4] = {FM25XXX_READ};
+
+	size_t addr_bytes = eeprom_fm25xxx_size_to_addr_bytes(config->size);
+	size_t op_len = 1 + addr_bytes;
+
+	switch (addr_bytes) {
+	case 1:
+		read_op[1] = offset & 0xff;
+		break;
+	case 2:
+		sys_put_be16(offset, &read_op[1]);
+		break;
+	case 3:
+		sys_put_be24(offset, &read_op[1]);
+		break;
+	default:
+		LOG_ERR("Invalid number of address bytes %u", addr_bytes);
+		break;
+	}
+
+	LOG_HEXDUMP_WRN(read_op, 4, "Read op");
+
+	const struct spi_buf tx_bufs[] = {{
+		.buf = &read_op,
+		.len = op_len,
+	}};
+	const struct spi_buf_set tx_buf_set = {
+		.buffers = tx_bufs,
+		.count = ARRAY_SIZE(tx_bufs),
+	};
+
+	const struct spi_buf rx_bufs[2] = {
+		{
+			.buf = NULL,
+			.len = op_len,
+		},
+		{
+			.buf = data,
+			.len = len,
+		},
+	};
+	const struct spi_buf_set rx_buf_set = {
+		.buffers = rx_bufs,
+		.count = ARRAY_SIZE(rx_bufs),
+	};
+
+	k_mutex_lock(&dev_data->lock, K_FOREVER);
+
+	ret = spi_transceive_dt(&config->spi, &tx_buf_set, &rx_buf_set);
+
+	k_mutex_unlock(&dev_data->lock);
+
+	if (ret != 0) {
+		LOG_ERR("Failed to read from FRAM");
+		return ret;
+	}
+
+	return 0;
+}
+
+int eeprom_fm25xxx_write(const struct device *dev, off_t offset, const void *data, size_t len)
+{
+	int ret;
+	const struct fm25xxx_config *config = dev->config;
+	struct fm25xxx_data *dev_data = dev->data;
+
+	if (config->readonly) {
+		LOG_ERR("Can not write to a readonly device");
+		return -EACCES;
+	}
+
+	if (offset + len > config->size) {
+		LOG_ERR("Can not write more data than the device size");
+		return -EINVAL;
+	}
+
+	if (len == 0) {
+		return 0;
+	}
+
+	uint8_t write_op[4] = {FM25XXX_WRITE};
+
+	size_t addr_bytes = eeprom_fm25xxx_size_to_addr_bytes(config->size);
+	size_t op_len = 1 + addr_bytes;
+
+	switch (addr_bytes) {
+	case 1:
+		write_op[1] = offset & 0xff;
+		break;
+	case 2:
+		sys_put_be16(offset, &write_op[1]);
+		break;
+	case 3:
+		sys_put_be24(offset, &write_op[1]);
+		break;
+	default:
+		LOG_ERR("Invalid number of address bytes %u", addr_bytes);
+		break;
```

Comment:
This should fail as it is an application programming error and it does not make sense to send out that command